samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -65,13 +65,7 @@
   private volatile String ephemeralPath = null;
   private final ZkKeyBuilder keyBuilder;
   private final int connectionTimeoutMs;
-  private ZkJobCoordinatorMetrics metrics;
-
-  public ZkUtils(ZkKeyBuilder zkKeyBuilder, ZkClient zkClient, int connectionTimeoutMs) {
-    this.keyBuilder = zkKeyBuilder;
-    this.connectionTimeoutMs = connectionTimeoutMs;
-    this.zkClient = zkClient;
-  }
+  private final ZkJobCoordinatorMetrics metrics;
 
   public ZkUtils(ZkKeyBuilder zkKeyBuilder, ZkClient zkClient, int connectionTimeoutMs, ZkJobCoordinatorMetrics metrics) {
```

Comment:
>Overloading it with ZkJobCoordinatorMetrics is confusing as this component - ZkUtils is also accessed from CoordinationService.

Just to be clear. ZKJobCoordinatorMetrics dependency into ZkUtils wasn’t introduced in this patch(as part of this change). It was how it was existent before. 

Just to fix a NPE, I didn’t want to get into refactoring stuff.

I agree with passing in just MetricsRegistry(that’s how it's done in rest of samza viz SamzaContainerMetrics, LocalStoreMonitorMetrics etc). 

Will make the change and update here.

> If you have more questions on why we are passing around MetricsRegistry and the

Thanks for the pointer, I will read about it. 

Though we achieve isolation in terms of registering/creating  metics and publishing metrics , following this pattern introduces dependency between different layers. 

Basic premise here is each layer expects the previous layer creating it to have a metricsRegistry(map) to pass it to current layer. There’s a fan out from top layer (who registers with reporters) to all lower level.  

In chain A-B-C if A registers metricsRegistry with reporters and B doesn’t need MetricsRegistry for reporting, it still has to have it as explicit dependency 
since C might require it to report metrics(which it should receive from B). 

If you already know the reasoning please share it with me(or poet out why this thinking is incorrect).